x-pack/plugins/actions/README.md

```diff
@@ -0,0 +1,177 @@
+|migrationVersion|The version of the most recent migrations. This is the same as `migrationVersion` in the saved objects API. See the saved objects API documentation.<br><br>In most cases, you can leave this empty.|object|
+
+#### `DELETE /api/action/{id}`: Delete action
+
+Params:
+
+|Property|Description|Type|
+|---|---|---|
+|id|The id of the action you're trying to delete.|string|
+
+#### `GET /api/action/_find`: Find actions
+
+Params:
+
+See the saved objects API documentation for find. All the properties are the same except that you cannot pass in `type`.
+
+#### `GET /api/action/{id}`: Get action
+
+Params:
+
+|Property|Description|Type|
+|---|---|---|
+|id|The id of the action you're trying to get.|string|
+
+#### `GET /api/action/types` List action types
+
+No parameters.
+
+#### `PUT /api/action/{id}`: Update action
+
+Params:
+
+|Property|Description|Type|
+|---|---|---|
+|id|The id of the action you're trying to update.|string|
+
+Payload:
+
+|Property|Description|Type|
+|---|---|---|
+|attributes.description|A description to reference and search in the future. This value will be used to populate dropdowns.|string|
+|attributes.actionTypeConfig|The configuration the action type expects. See related action type to see what attributes is expected. This will also validate against the action type if config validation is defined.|object|
+|references|An array of `name`, `type` and `id`. This is the same as `references` in the saved objects API. See the saved objects API documentation.<br><br>In most cases, you can leave this empty.|Array|
+|version|The document version when read|string|
+
+## Firing actions
+
+The plugin exposes a fire function that you can use to fire actions.
+
+**server.plugins.actions.fire(options)**
+
+The following table describes the properties of the `options` object.
+
+|Property|Description|Type|
+|---|---|---|
+|id|The id of the action you want to fire.|string|
+|params|The `params` value to give the action type executor.|object|
+|namespace|The saved object namespace the action exists within.|string|
+|basePath|This is a temporary parameter, but we need to capture and track the value of `request.getBasePath()` until future changes are made.<br><br>In most cases this can be `undefined` unless you need cross spaces support.|string|
+
+### Example
+
+This example makes action `3c5b2bd4-5424-4e4b-8cf5-c0a58c762cc5` fire an email. The action plugin will load the saved object and find what action type to call with `params`.
+
+```
+server.plugins.actions.fire({
+  id: '3c5b2bd4-5424-4e4b-8cf5-c0a58c762cc5',
+  params: {
+    from: 'example@elastic.co',
+    to: ['destination@elastic.co'],
+    subject: 'My email subject',
+    body: 'My email body',
+  },
+  namespace: undefined, // The namespace the action exists within
+  basePath: undefined, // Usually `request.getBasePath();` or `undefined`
+});
+```
```

Comment:
Reading through this document I have a good understanding of the different pieces like action types, CRUD for actions, and firing.  But an example that shows the full usage from creation of an action to the action being completed might help answer some questions. For example: after I create an action and fire it,  how do I get its status - how do I know it worked?  Do I get that through task manager, do I poll the action to get a status?

Reply:
> For example: after I create an action and fire it, how do I get its status - how do I know it worked? Do I get that through task manager, do I poll the action to get a status?

TBD; seems like we'll have some kind of history &| audit log of actions run, so you'd need to poll that.  We'll need to have the action fire return some kind of id for the action fire status, that would eventually show up in the history &| audit log.

Reply:
Another question related to action lifecycle, once an action has been fired and executed,  should I delete it or keep it for history?  Are there any consequences to not deleting them, do they get cleaned up periodically?

Reply:
> do they get cleaned up periodically?

I just did a test regarding this, if a task isn't re-scheduled, it gets removed from the task manager. The task manager periodically cleans up expired tasks.

Reply:
I just had a conversation with @bmcconaghy about this. We're going to add a REST API to fire actions (PR coming soon). We will change the behaviour of this manual function and make it hold off until the execution is complete. This function will then be able to throw errors, if any occur, during the execution.